Source/Initialization/WarpXInitData.cpp

```diff
@@ -695,46 +695,172 @@ WarpX::InitLevelData (int lev, Real /*time*/)
+        Store_parserString(pp, "My_external_grid_function(x,y,z)",
                                                     str_My_ext_grid_function);
-       Store_parserString(pp, "Mz_external_grid_function(x,y,z)",
+        Store_parserString(pp, "Mz_external_grid_function(x,y,z)",
                                                     str_Mz_ext_grid_function);
 
-       Mxfield_parser.reset(new ParserWrapper<3>(
-                         makeParser(str_Mx_ext_grid_function,{"x","y","z"})));
-       Myfield_parser.reset(new ParserWrapper<3>(
-                         makeParser(str_My_ext_grid_function,{"x","y","z"})));
-       Mzfield_parser.reset(new ParserWrapper<3>(
-                         makeParser(str_Mz_ext_grid_function,{"x","y","z"})));
-
-       // Initialize Mfield_fp with external function
-       InitializeExternalFieldsOnGridUsingParser(Mfield_fp[lev][0].get(),
-                                                 Mfield_fp[lev][1].get(),
-                                                 Mfield_fp[lev][2].get(),
-                                                 getParser(Mxfield_parser),
-                                                 getParser(Myfield_parser),
-                                                 getParser(Mzfield_parser),
-                                                 lev);
-       if (lev > 0) {
-          InitializeExternalFieldsOnGridUsingParser(Mfield_aux[lev][0].get(),
-                                                    Mfield_aux[lev][1].get(),
-                                                    Mfield_aux[lev][2].get(),
-                                                    getParser(Mxfield_parser),
-                                                    getParser(Myfield_parser),
-                                                    getParser(Mzfield_parser),
-                                                    lev);
+        Mxfield_parser.reset(new ParserWrapper<3>(
+                                 makeParser(str_Mx_ext_grid_function,{"x","y","z"})));
+        Myfield_parser.reset(new ParserWrapper<3>(
+                                 makeParser(str_My_ext_grid_function,{"x","y","z"})));
+        Mzfield_parser.reset(new ParserWrapper<3>(
+                                 makeParser(str_Mz_ext_grid_function,{"x","y","z"})));
+
+        {   // use this brace so Mx, My, Mz go out of scope
+            // we need 1 more ghost cell than Mfield_fp has because
+            // we are averaging to faces, including the ghost faces
+            BoxArray bx = Mfield_fp[lev][0]->boxArray();
+            amrex::MultiFab Mx(bx.enclosedCells(), Mfield_fp[lev][0]->DistributionMap(), 1, Mfield_fp[lev][0]->nGrow()+1);
+            amrex::MultiFab My(bx.enclosedCells(), Mfield_fp[lev][1]->DistributionMap(), 1, Mfield_fp[lev][1]->nGrow()+1);
+            amrex::MultiFab Mz(bx.enclosedCells(), Mfield_fp[lev][2]->DistributionMap(), 1, Mfield_fp[lev][2]->nGrow()+1);
```

Comment:
would it be the same if we do 
`Mx(Mfield_fp[lev][0]->boxArray.enclosedCells(), ..... , ....., ...)`

Reply:
I tried that but amrex doesn't seem to like it for some reason.  Maybe it's because boxArray() returns a const BoxArray and therefore you can't modify it to be cell-centered with enclosedCells all in one line?

Reply:
oh thank you for that explanation.